// vuetify/lib/components/VChipGroup/VChipGroup.js
var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

// Styles
import '../../../src/components/VChipGroup/VChipGroup.sass';
// Extensions
import { BaseSlideGroup } from '../VSlideGroup/VSlideGroup';
// Mixins
import Colorable from '../../mixins/colorable';
// Utilities
import mixins from '../../util/mixins';
/* @vue/component */
export default mixins(BaseSlideGroup, Colorable).extend({
    name: 'v-chip-group',
    provide: function provide() {
        return {
            chipGroup: this
        };
    },

    props: {
        choice: Boolean,
        column: Boolean,
        outlined: Boolean
    },
    computed: {
        classes: function classes() {
            return _extends({}, BaseSlideGroup.options.computed.classes.call(this), {
                'v-chip-group': true,
                'v-chip-group--column': this.column
            });
        }
    },
    watch: {
        column: function column(val) {
            if (val) this.scrollOffset = 0;
            this.$nextTick(this.onResize);
        }
    },
    methods: {
        genData: function genData() {
            return this.setTextColor(this.color, _extends({}, BaseSlideGroup.options.methods.genData.call(this)));
        }
    }
});
