// vuetify/es5/components/VList/VListItemGroup.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; // Styles

// Extensions

// Mixins

// Utilities


require('../../../src/components/VList/VListItemGroup.sass');

var _VItemGroup = require('../VItemGroup/VItemGroup');

var _colorable = require('../../mixins/colorable');

var _colorable2 = _interopRequireDefault(_colorable);

var _mixins = require('../../util/mixins');

var _mixins2 = _interopRequireDefault(_mixins);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = (0, _mixins2.default)(_VItemGroup.BaseItemGroup, _colorable2.default).extend({
    name: 'v-list-item-group',
    provide: function provide() {
        return {
            isInGroup: true,
            listItemGroup: this
        };
    },

    computed: {
        classes: function classes() {
            return _extends({}, _VItemGroup.BaseItemGroup.options.computed.classes.call(this), {
                'v-list-item-group': true
            });
        }
    },
    methods: {
        genData: function genData() {
            return this.setTextColor(this.color, _extends({}, _VItemGroup.BaseItemGroup.options.methods.genData.call(this), {
                attrs: {
                    role: 'listbox'
                }
            }));
        }
    }
});
