// vuetify/es5/components/VList/VListItemAvatar.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; // Components


var _VAvatar = require('../VAvatar');

var _VAvatar2 = _interopRequireDefault(_VAvatar);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* @vue/component */
exports.default = _VAvatar2.default.extend({
    name: 'v-list-item-avatar',
    props: {
        horizontal: Boolean,
        size: {
            type: [Number, String],
            default: 40
        }
    },
    computed: {
        classes: function classes() {
            return _extends({
                'v-list-item__avatar--horizontal': this.horizontal
            }, _VAvatar2.default.options.computed.classes.call(this), {
                'v-avatar--tile': this.tile || this.horizontal
            });
        }
    },
    render: function render(h) {
        var render = _VAvatar2.default.options.render.call(this, h);
        render.data = render.data || {};
        render.data.staticClass += ' v-list-item__avatar';
        return render;
    }
});
