// vuetify/lib/services/goto/easing-patterns.js
// linear
export var linear = function linear(t) {
  return t;
};
// accelerating from zero velocity
export var easeInQuad = function easeInQuad(t) {
  return Math.pow(t, 2);
};
// decelerating to zero velocity
export var easeOutQuad = function easeOutQuad(t) {
  return t * (2 - t);
};
// acceleration until halfway, then deceleration
export var easeInOutQuad = function easeInOutQuad(t) {
  return t < 0.5 ? 2 * Math.pow(t, 2) : -1 + (4 - 2 * t) * t;
};
// accelerating from zero velocity
export var easeInCubic = function easeInCubic(t) {
  return Math.pow(t, 3);
};
// decelerating to zero velocity
export var easeOutCubic = function easeOutCubic(t) {
  return Math.pow(--t, 3) + 1;
};
// acceleration until halfway, then deceleration
export var easeInOutCubic = function easeInOutCubic(t) {
  return t < 0.5 ? 4 * Math.pow(t, 3) : (t - 1) * (2 * t - 2) * (2 * t - 2) + 1;
};
// accelerating from zero velocity
export var easeInQuart = function easeInQuart(t) {
  return Math.pow(t, 4);
};
// decelerating to zero velocity
export var easeOutQuart = function easeOutQuart(t) {
  return 1 - Math.pow(--t, 4);
};
// acceleration until halfway, then deceleration
export var easeInOutQuart = function easeInOutQuart(t) {
  return t < 0.5 ? 8 * t * t * t * t : 1 - 8 * --t * t * t * t;
};
// accelerating from zero velocity
export var easeInQuint = function easeInQuint(t) {
  return Math.pow(t, 5);
};
// decelerating to zero velocity
export var easeOutQuint = function easeOutQuint(t) {
  return 1 + Math.pow(--t, 5);
};
// acceleration until halfway, then deceleration
export var easeInOutQuint = function easeInOutQuint(t) {
  return t < 0.5 ? 16 * Math.pow(t, 5) : 1 + 16 * Math.pow(--t, 5);
};
