// vuetify/es5/components/VBtnToggle/VBtnToggle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; // Styles

// Mixins

// Utilities


require('../../../src/components/VBtnToggle/VBtnToggle.sass');

var _buttonGroup = require('../../mixins/button-group');

var _buttonGroup2 = _interopRequireDefault(_buttonGroup);

var _colorable = require('../../mixins/colorable');

var _colorable2 = _interopRequireDefault(_colorable);

var _mixins = require('../../util/mixins');

var _mixins2 = _interopRequireDefault(_mixins);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* @vue/component */
exports.default = (0, _mixins2.default)(_buttonGroup2.default, _colorable2.default).extend({
    name: 'v-btn-toggle',
    props: {
        rounded: Boolean
    },
    computed: {
        classes: function classes() {
            return _extends({}, _buttonGroup2.default.options.computed.classes.call(this), {
                'v-btn-toggle': true,
                'v-btn-toggle--rounded': this.rounded
            }, this.themeClasses);
        }
    },
    methods: {
        genData: function genData() {
            return this.setTextColor(this.color, _extends({}, _buttonGroup2.default.options.methods.genData.call(this)));
        }
    }
});
