// vuetify/es5/components/VBottomSheet/VBottomSheet.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };
// Extensions


require('../../../src/components/VBottomSheet/VBottomSheet.sass');

var _VDialog = require('../VDialog/VDialog');

var _VDialog2 = _interopRequireDefault(_VDialog);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* @vue/component */
exports.default = _VDialog2.default.extend({
    name: 'v-bottom-sheet',
    props: {
        inset: Boolean,
        maxWidth: {
            type: [String, Number],
            default: 'auto'
        },
        transition: {
            type: String,
            default: 'bottom-sheet-transition'
        }
    },
    computed: {
        classes: function classes() {
            return _extends({}, _VDialog2.default.options.computed.classes.call(this), {
                'v-bottom-sheet': true,
                'v-bottom-sheet--inset': this.inset
            });
        }
    }
});
