// vuetify/es5/services/application/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.Application = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _service = require('../service');

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } // Extensions


var Application = exports.Application = function (_Service) {
    _inherits(Application, _Service);

    function Application() {
        _classCallCheck(this, Application);

        var _this = _possibleConstructorReturn(this, (Application.__proto__ || Object.getPrototypeOf(Application)).apply(this, arguments));

        _this.bar = 0;
        _this.top = 0;
        _this.left = 0;
        _this.insetFooter = 0;
        _this.right = 0;
        _this.bottom = 0;
        _this.footer = 0;
        _this.application = {
            bar: {},
            top: {},
            left: {},
            insetFooter: {},
            right: {},
            bottom: {},
            footer: {}
        };
        return _this;
    }

    _createClass(Application, [{
        key: 'register',
        value: function register(uid, location, size) {
            this.application[location][uid] = size;
            this.update(location);
        }
    }, {
        key: 'unregister',
        value: function unregister(uid, location) {
            if (this.application[location][uid] == null) return;
            delete this.application[location][uid];
            this.update(location);
        }
    }, {
        key: 'update',
        value: function update(location) {
            this[location] = Object.values(this.application[location]).reduce(function (acc, cur) {
                return acc + cur;
            }, 0);
        }
    }]);

    return Application;
}(_service.Service);

Application.property = 'application';
