// vuetify/lib/components/VList/index.js
import { createSimpleFunctional } from '../../util/helpers';
import VList from './VList';
import VListGroup from './VListGroup';
import VListItem from './VListItem';
import VListItemGroup from './VListItemGroup';
import VListItemAction from './VListItemAction';
import VListItemAvatar from './VListItemAvatar';
import VListItemIcon from './VListItemIcon';
export var VListItemActionText = createSimpleFunctional('v-list-item__action-text', 'span');
export var VListItemContent = createSimpleFunctional('v-list-item__content', 'div');
export var VListItemTitle = createSimpleFunctional('v-list-item__title', 'div');
export var VListItemSubtitle = createSimpleFunctional('v-list-item__subtitle', 'div');
export { VList, VListGroup, VListItem, VListItemAction, VListItemAvatar, VListItemIcon, VListItemGroup };
export default {
    $_vuetify_subcomponents: {
        VList: VList,
        VListGroup: VListGroup,
        VListItem: VListItem,
        VListItemAction: VListItemAction,
        VListItemActionText: VListItemActionText,
        VListItemAvatar: VListItemAvatar,
        VListItemContent: VListItemContent,
        VListItemGroup: VListItemGroup,
        VListItemSubtitle: VListItemSubtitle,
        VListItemTitle: VListItemTitle
    }
};
