// vuetify/es5/components/VDataTable/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.VMobileRow = exports.VVirtualTable = exports.VSimpleTable = exports.VDataTableHeader = exports.VRowGroup = exports.VRow = exports.VTableOverflow = exports.VEditDialog = exports.VDataTable = undefined;

var _helpers = require('../../util/helpers');

var _VDataTable = require('./VDataTable');

var _VDataTable2 = _interopRequireDefault(_VDataTable);

var _VDataTableHeader = require('./VDataTableHeader');

var _VDataTableHeader2 = _interopRequireDefault(_VDataTableHeader);

var _VRow = require('./VRow');

var _VRow2 = _interopRequireDefault(_VRow);

var _VRowGroup = require('./VRowGroup');

var _VRowGroup2 = _interopRequireDefault(_VRowGroup);

var _VEditDialog = require('./VEditDialog');

var _VEditDialog2 = _interopRequireDefault(_VEditDialog);

var _VSimpleTable = require('./VSimpleTable');

var _VSimpleTable2 = _interopRequireDefault(_VSimpleTable);

var _VVirtualTable = require('./VVirtualTable');

var _VVirtualTable2 = _interopRequireDefault(_VVirtualTable);

var _VMobileRow = require('./VMobileRow');

var _VMobileRow2 = _interopRequireDefault(_VMobileRow);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var VTableOverflow = (0, _helpers.createSimpleFunctional)('v-table__overflow');
exports.VDataTable = _VDataTable2.default;
exports.VEditDialog = _VEditDialog2.default;
exports.VTableOverflow = VTableOverflow;
exports.VRow = _VRow2.default;
exports.VRowGroup = _VRowGroup2.default;
exports.VDataTableHeader = _VDataTableHeader2.default;
exports.VSimpleTable = _VSimpleTable2.default;
exports.VVirtualTable = _VVirtualTable2.default;
exports.VMobileRow = _VMobileRow2.default;
exports.default = {
    $_vuetify_subcomponents: {
        VDataTable: _VDataTable2.default,
        VDataTableHeader: _VDataTableHeader2.default,
        VRow: _VRow2.default,
        VMobileRow: _VMobileRow2.default,
        VRowGroup: _VRowGroup2.default,
        VEditDialog: _VEditDialog2.default,
        VTableOverflow: VTableOverflow,
        VSimpleTable: _VSimpleTable2.default,
        VVirtualTable: _VVirtualTable2.default
    }
};
